### PR TITLE
Use longs in the rate limiter and handle negative adjustments.

### PR DESCRIPTION
**Pull request by hsaraogi**
**Priority (whenever / two weeks / yesterday)**: soon.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2758)

### CODE REVIEW

File: qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/guava/SmoothRateLimiter.java [hunk 1]
```diff
@@ -382,20 +383,31 @@ final long reserveEarliestAvailable(int requiredPermits, long nowMicros) {
...
-                    storedPermits
-                            + (nowMicros - nextFreeTicketMicros) / coolDownIntervalMicros());
-            nextFreeTicketMicros = nowMicros;
+        if (nowNanos > nextFreeTicketNanos) {
+            storedPermits = min(maxPermits, storedPermits + (nowNanos - nextFreeTicketNanos) / coolDownIntervalNanos());
+            nextFreeTicketNanos = nowNanos;
+        }
+    }
+
+    void returnPermitsConsumedEarlier(double permitsToReturn, long nowNanos) {
```
Comment (nziebart) [inline qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/guava/SmoothRateLimiter.java]: can we sync in person on how this works?

File: qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/guava/RateLimiter.java [hunk 2]
```diff
@@ -235,11 +237,11 @@ public final void setRate(double permitsPerSecond) {
         checkArgument(
                 permitsPerSecond > 0.0 && !Double.isNaN(permitsPerSecond), "rate must be positive");
         synchronized (mutex()) {
-            doSetRate(permitsPerSecond, stopwatch.readMicros());
+            doSetRate(permitsPerSecond, stopwatch.readNanos());
```
Comment (nziebart) [inline qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/guava/RateLimiter.java]: why the change from micros to nanos? it's generally nicer, but without a strong need we should prefer avoiding changes to this copied code

File: qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/QosRateLimiter.java [hunk 3]
```diff
@@ -123,9 +123,14 @@ private synchronized void createRateLimiterAtomically() {
      */
     public void recordAdjustment(long adjustmentUnits) {
         if (adjustmentUnits > 0) {
-            rateLimiter.steal(Ints.saturatedCast(adjustmentUnits)); // TODO(nziebart): deal with longs
+            rateLimiter.steal(adjustmentUnits);
+        } else {
+            try {
+                rateLimiter.returnPermits(LongMath.checkedMultiply(adjustmentUnits, -1), stopwatch.readNanos());
+            } catch (ArithmeticException e) {
```
Comment (nziebart) [inline qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/QosRateLimiter.java]: can multiplying by -1 ever cause an issue?

File: qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/QosRateLimiter.java [hunk 4]
```diff
@@ -123,9 +123,14 @@ private synchronized void createRateLimiterAtomically() {
      */
     public void recordAdjustment(long adjustmentUnits) {
         if (adjustmentUnits > 0) {
-            rateLimiter.steal(Ints.saturatedCast(adjustmentUnits)); // TODO(nziebart): deal with longs
+            rateLimiter.steal(adjustmentUnits);
+        } else {
+            try {
+                rateLimiter.returnPermits(LongMath.checkedMultiply(adjustmentUnits, -1), stopwatch.readNanos());
```
Comment (nziebart) [inline qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/QosRateLimiter.java]: nit: let's name this `return` for consistency with the other method names

File: qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/guava/RateLimiter.java (hunk 2 above)
Comment (hsaraogi) [reply qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/guava/RateLimiter.java]: If you have longs you might often need to wait via sleeping for just a few nanos, as sleeping in micros would be excessive.

File: qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/guava/SmoothRateLimiter.java (hunk 1 above)
Comment (hsaraogi) [reply qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/guava/SmoothRateLimiter.java]: Done.

File: qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/guava/RateLimiter.java (hunk 2 above)
Comment (hsaraogi) [reply qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/guava/RateLimiter.java]: If you have longs you might often need to be sleeping for just a few nanos, as the `stableand sleeping in micros would be non-performant.

File: qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/QosRateLimiter.java (hunk 3 above)
Comment (hsaraogi) [reply qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/QosRateLimiter.java]: Yup, if its Long.MIN_VALUE, caught by a test. :D

File: qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/QosRateLimiter.java (hunk 4 above)
Comment (hsaraogi) [reply qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/QosRateLimiter.java]: LOL, `return` is a keyword.

File: qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/guava/SmoothRateLimiter.java (hunk 1 above)
Comment (nziebart) [reply qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/guava/SmoothRateLimiter.java]: this method should have a changelog

File: qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/QosRateLimiter.java (hunk 3 above)
Comment (nziebart) [reply qos-service-impl/src/main/java/com/palantir/atlasdb/qos/ratelimit/QosRateLimiter.java]: nit: move the safe multiply operation to its own method (e.g.`safeAbs`) so the `try/catch` isn't around the whole method call

**Comment by nziebart** [comment]
@hsaraogi i went ahead and addressed my own comments if you want to take a last look
